Mothership Rework (for real, thiss time)

SirChuckOfTheChuckles opened the pull request:
Generator PR Link: https://github.com/Ziktofel/Archipelago/pull/386

### Overview
Implements two other Mothership variants, the Aiur Mothership and the Purifier Mothership. These two variants are morphed from their respective Mothership Cores, which can be created from the Nexus when you have a Cybernetics Core. They require a Fleet Beacon to be morphed from a Mothership Core.

![image](https://github.com/user-attachments/assets/6640303c-0871-471e-915f-2aa8c001cf34)

In-game, the units themselves are simply named "Mothership Core" or "Mothership", but they have a subtitle indicating which faction they belong to, in order to make the names not super long.

![image](https://github.com/user-attachments/assets/15ec54eb-5111-400f-bd5b-38cf9416cf5e)


### Mothership Cores

The Aiur and Purifier Mothership Cores are similar, but each has one unique ability and a unique attack.
**Time Warp** Causes enemies' attack and move speed to be slowed by 50% while they stay in the target area. Lasts 10 seconds.

https://github.com/user-attachments/assets/3a4871bb-4940-4cb3-bc4d-b940a39c7cc8

**Nexus Recall** Teleports all nearby units and the Mothership to the nearby Nexus.

https://github.com/user-attachments/assets/f3a69037-ffc2-4d52-8151-e26f58b04b48

**Morph to Mothership** Morphs the Mothership Core into their respective Mothership. They cannot take any actions during this time, but the process can be cancelled.

https://github.com/user-attachments/assets/36a20be6-5b81-49ce-bfae-c1a8e2f8a624
(Note, morph time is heavily sped up in the video).

**Battery Overcharge** Unique to Aiur Mothership Core. Energizes a shield battery to grant 100% increased heal speed and cause it to not consume energy when restoring shields.


https://github.com/user-attachments/assets/624e3d73-fc8c-487e-9372-ccb110465283


**Power Redirection** Unique to Purifier Mothership Core. Transfer power from the Mothership Core to a Photon or Khaydarin Monolith, powering the building for the duration while granting combat bonuses (attack speed and additional targets hit for photon cannon, splash damage and attack speed for khaydarin monolith).


https://github.com/user-attachments/assets/6bf43466-3d53-4353-b39b-bac99cdf46cd

https://github.com/user-attachments/assets/f26b4b95-bfba-4549-b149-58445ade7f7f


### Motherships

Mothership variants have 3 submenus for their abilities/upgrades: Offensive abilities, such as Thermal Lance or Planet Cracker; Utility abilities, such as Black Hole or Mass Recall, and Auxiliary Upgrades, which mainly buff the stats of the Mothership or slightly change the effect of it's primary weapons. Offensive abilities are unique across all Motherships, Utility abilities are the same (with the Tal'darim Mothership's utility being slightly better, shown later), and Auxiliary is a mix between unique and the same. The abilities/upgrades for a mothership do NOT persist through death (aside from some auxiliary upgrades, which it will state in it's description).

https://github.com/user-attachments/assets/2a2a3ac2-2b14-4361-a167-2bc312d3b67c



**Aiur Mothership** Provides a cloaking aura to all units around it, and does a massive slow on it's attack (war council upgrade). Can use Revitalize Shields and Astral Restoration

Revitalize Shields (50 Energy, No CD): Restores 60 shields to friendly units and structures in the target area.

https://github.com/user-attachments/assets/ad52f64b-7b53-4d12-a9cf-83bb5f1f7b74

Astral Restoration (100 Energy, 240 sec CD) Increase the radius of Cloaking Field by 50%, and heal 160 shields and 160 life to all units and structures within radius of the cloaking field over 10 seconds.

https://github.com/user-attachments/assets/5e8aafdc-cc4a-4475-a262-b476e1c715b5

Recursive Photons (Unique Auxiliary Upgrade; 250 Minerals 250 Gas): The primary weapon of the Mothership will now continously deal damage to the target as long as the Mothership is in range of them


https://github.com/user-attachments/assets/9a85f107-d073-4e6d-adbb-1963435c4d92



**Purifier Mothership** Provides pylon power to units and structures around it (war council upgrade), and attacks can hit up 3 targets within range

https://github.com/user-attachments/assets/02297bcd-b7dc-4200-be00-07c1e4b49cc8

Overload Weapon Systems (75 energy, No CD) Disable the movement of the Mothership, but allow it to hit 3 additional targets with it's attacks and grant it 25% attack speed for 10 seconds. After this effect ends, stun the Mothership for 2 seconds.


https://github.com/user-attachments/assets/b9c6108f-5c35-49dc-a8e6-40616683505a


Planet Cracker (150 Energy, 360 sec CD) Teleport to a nearby location and fire the Planet Cracker, dealing 720 damage to **ALL** targets in the radius over 10 seconds. Radius increases as the duration goes on.

https://github.com/user-attachments/assets/cda5fae9-5a26-4ef8-833f-df23b0692ec7


Purified Embers (Unique Auxiliary Upgrade; 200 Minerals, 200 Gas): Deals damage over time to enemies attacked by the purifier mothership's primary weapon.

**Tal'darim Mothership** The Tal'darim mothership can install Thermal and Black Hole for free, to perserve vanilla functionality. It also gains unique bonuses for the Utility abilties (increased radius, minor shielding/damage increase), so Aiur and Purifier mothership can still have an option like black hole without it being too powerful. 


https://github.com/user-attachments/assets/d31bd57c-e940-4d1f-8096-9f5619764433


Summon Death Fleet (125 Energy, 300 sec CD) Summons a group of 6 Destroyers to control with a timed life of 30 seconds.


https://github.com/user-attachments/assets/f381d47c-f442-4401-a329-cf44501d4d26

Hunger For Slaughter (Unique Auxiliary Upgrade; 400 Minerals, 400 Gas): The Tal'darim Mothership's primary weapon gains 100% attack speed and deals damage much quicker, but has it's base damage significantly reduced.


https://github.com/user-attachments/assets/390422d7-a389-438e-bc83-971841f941e8


### Auxiliary Upgrade Overview
Each Mothership has baseline 3 auxiliary upgrades that stack up to 10 times. Once a single upgrade is researched 10 times, it becomes a permanent upgrade on that mothership, persisting through death and granting the Mothership an additional bonus. To clarify, researching an upgrade like Refined Weapons on a Purifier Mothership will NOT research it on it's aiur or tal'darim variants; each Mothership has separate versions of these upgrades.
These upgrades are:
Refined Weapons: Increase primary weapon damage by X. After upgrading 10 times, damage becomes permanent and gain 15% attack speed.
Bolstered Plating: Increase life and shields by X. After upgrading X times, life and shield increase becomes permanent also gain 1 life and shield armor.
Optimized Energy Usage: Increase energy regen by X. After upgrading 10 times, energy regen bonus becomes  permanent, and gain 0.5 move speed.

![image](https://github.com/user-attachments/assets/8c22e8fa-085e-4adb-a469-b943212dac22)


### Clarifications on baseline functionality/items
Every mothership in a run will begin with at least one offensive ability (thermal lance, overload weapon systems, or revitalize shields), one utility ability (black hole), and the three stacking auxiliary upgrades listed above. 
![image](https://github.com/user-attachments/assets/4ed1a63e-b685-4c30-96a8-0a784fa5924d)

MatthewMarinets commented on Mods/ArchipelagoPlayer.SC2Mod/deleteme.txt:
Not sure why this file was added here.

SirChuckOfTheChuckles replied:
was a mistake I missed, thanks for the catch

Ziktofel commented:
![Snímek obrazovky pořízený 2025-01-15 21-24-28](https://github.com/user-attachments/assets/ff0a8591-85b7-4f61-a4ff-a4f7f16d32f3)
Tal'Darim ones should have the abilities installed by default as they're directly available in vanilla game (and are free anyways)
![Snímek obrazovky pořízený 2025-01-15 21-26-10](https://github.com/user-attachments/assets/620cf584-df48-4aca-ba87-f55f7a845cd9)
Auxilliary upgrades for Tal'Darim Mothership aren't part of the vanilla game, should be gated behind an item (I can purchase it outright without having any of the new upgrades and those aren't blocked in cleanProtossTech function in Triggers). Either way this submenu shall be sticky

Ziktofel commented:
![Snímek obrazovky pořízený 2025-01-15 21-34-23](https://github.com/user-attachments/assets/3ff36ed9-4dfe-4574-8c4f-a6a5eb05e7b6)
Behavior string

Ziktofel commented:
![Snímek obrazovky pořízený 2025-01-15 21-43-02](https://github.com/user-attachments/assets/019fcb9f-4ebf-49f8-a2c4-946a0102e8f5)
Bug: Doesn't replace with the maxed upgrade like other motherships have. Can dump extra resources without any effect.

Ziktofel commented:
![Snímek obrazovky pořízený 2025-01-15 21-57-04](https://github.com/user-attachments/assets/3bb37f0b-d161-444d-b08f-14bdb74bba17)
Bombers says disabled after installing them

Ziktofel commented:
I'd recommend to be able to swap the ability of the mothership (just paying the fee for new one, losing the old one). Tal'darim should then start with the vanilla ones

Ziktofel commented on Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/UpgradeData.xml:
```diff
@@ -14525,11 +14612,130 @@
+        <EditorCategories value="Race:Protoss,UpgradeType:Talents"/>
+        <AffectedUnitArray value="AP_MothershipAiur"/>
+    </CUpgrade>
+    <CUpgrade id="AP_HungerForSlaughter">
+        <EffectArray Reference="Weapon,AP_SOAMothershipv4Weapon,RateMultiplier" Value="1"/>
+        <EffectArray Operation="Set" Reference="Weapon,AP_SOAMothershipv4Weapon,Effect" Value="AP_SOAMothershipv4FastBeamSet"/>
+        <EffectArray Operation="Subtract" Reference="Effect,AP_SOAMothershipv4BeamDamage,Amount" Value="19"/>
+        <EditorCategories value="Race:Protoss,UpgradeType:Talents"/>
+        <AffectedUnitArray value="AP_SOAMothershipv4"/>
+        <AffectedUnitArray value="AP_MothershipTaldarim"/>
+    </CUpgrade>
```
An upgrade received by an item shouldn't have a drawback. This could be solved by a toggleable firing mode driven by a Behavior. Check Infested Libs that do that

Ziktofel replied:
Still missing weapon toggle

Ziktofel commented:
Planet Cracker hurts even own Bombers, is that intended?

Ziktofel commented:
![Snímek obrazovky pořízený 2025-01-15 22-23-37](https://github.com/user-attachments/assets/98f3d5d7-0885-4dfb-a4be-2d3e7090a540)
Bug: Can train another Purifier mothership core if I have a Purifier Mothership deployed (I have the mobile power field upgrade on)

Ziktofel commented:
![Snímek obrazovky pořízený 2025-01-15 22-27-39](https://github.com/user-attachments/assets/1855d5af-2ed1-4bfb-b349-d012c819c152)
Really only 5 energy?

Ziktofel commented on Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/BehaviorData.xml:
```diff
@@ -4987,6 +5035,381 @@
+        <MaxStackCount value="10"/>
+        <Modification>
+            <VitalMaxArray index="Life" value="50"/>
+            <VitalMaxArray index="Shields" value="50"/>]
```
Excess text node `]`

Ziktofel commented on Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/BehaviorData.xml:
```diff
@@ -4987,6 +5035,381 @@
+        <MaxStackCount value="10"/>
+        <Modification>
+            <VitalMaxArray index="Life" value="40"/>
+            <VitalMaxArray index="Shields" value="40"/>]
```
Cleanup the `]` symbol

Ziktofel commented on Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/BehaviorData.xml:
```diff
@@ -4987,6 +5035,381 @@
+        <MaxStackCount value="10"/>
+        <Modification>
+            <VitalMaxArray index="Life" value="100"/>
+            <VitalMaxArray index="Shields" value="100"/>]
```
Cleanup the `]` symbol

Ziktofel commented:
![Snímek obrazovky pořízený 2025-01-15 22-34-36](https://github.com/user-attachments/assets/a7ccf4de-62f2-47d5-b581-9783d0f2427d)
Purifier with Mobile Power Field: Missing the armor/shield upgrade level numbers

Ziktofel commented on Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/ButtonData.xml:
```diff
@@ -4677,6 +5110,11 @@
+    <CButton id="AP_CancelMothershipMorph">
+        <Icon value="Assets\Textures\btn-command-cancel.dds"/>
+        <AlertIcon value="Assets\Textures\btn-command-cancel.dds"/>
+        <EditorCategories value="Race:Protoss"/>
```
Should be race-tinted

Ziktofel commented on Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/EffectData.xml:
```diff
@@ -8053,7 +8053,7 @@
         <EditorCategories value="Race:Protoss"/>
         <Visibility value="Visible"/>
-        <Kind value="Ranged"/>
+        <Kind value="Melee"/>
```
Why this change? Shall Interceptors by any means be effective under Dark Swarm?

Ziktofel commented on Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/EffectData.xml:
```diff
@@ -17196,6 +18037,211 @@
+        <EffectArray value="AP_MothershipCoreNexusRecallApplyBehavior"/>
+    </CEffectSet>
+    <CEffectApplyBehavior id="AP_MothershipCoreNexusRecallApplyBehavior">
+        <ValidatorArray value="NotReleasedInterceptor"/>
+        <ValidatorArray index="0" value="AP_NotAPLarvaEgg"/>
+        <ValidatorArray value="AP_NotAPLarva"/>
+        <ValidatorArray value="AP_NotAbducted"/>
```
The one with `index="0"` shall be first

Ziktofel commented on Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/EffectData.xml:
```diff
@@ -17196,6 +18037,211 @@
+        <Behavior value="AP_MothershipRecallTaldarimDummy"/>
+    </CEffectApplyBehavior>
+    <CEffectApplyBehavior id="AP_MothershipNexusRecallTaldarimAB">
+        <ValidatorArray value="NotReleasedInterceptor"/>
+        <ValidatorArray index="0" value="AP_NotAPLarvaEgg"/>
+        <ValidatorArray value="AP_NotAPLarva"/>
+        <ValidatorArray value="AP_NotAbducted"/>
+        <ValidatorArray value="AP_CasterIsTaldarimMothership"/>
```
The one with `index="0"` shall be first

Ziktofel commented on Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/EffectData.xml:
```diff
@@ -18156,7 +19277,7 @@
     </CEffectSet>
     <CEffectDamage id="AP_CarrierPurifierBomberDamage" parent="DU_WEAP">
         <EditorCategories value="Race:Terran"/>
-        <Kind value="Ranged"/>
+        <Kind value="Splash"/>
```
That means effective under Dark Swarm, I think it should be this way

SirChuckOfTheChuckles commented:
Made following changes:
Fixed issue where some behaviors were missing names/tooltips
Fixed issue where paths for level 4 and 5 W/A upgrades were incorrect
Fixed issue where some triggers were not working
Adjusted costs for some auxiliary upgrades/abilities
Mothership Purifier Bombers now recall back to the Mothership when it casts Planet Cracker (they can still die at the start of the ability, but I'm willing to call this intended design)
Fixed an issue where the Purifier Mothership was missing shield/armor effects when the power field item was collected
Fixed an issue where interceptors/bombers were not classified as range
--- I Initially made this change to make refined weapons work, thinking it didn't effect anything while not realizing it actually did. Made decoy display effects instead.
Mothership hangar slots don't appear unless hangars are installed
Hangar upgrades now persist through death
Mothership abilities can now be swapped, but the installation cost must be paid whenever a swap is made.
Tal'darim Motherships start with Thermal Lance and Black Hole installed.
Combined Nexus Recall and Mass Recall item upgrades for the Mothership. Created an item which unlocks the 3 repeatable upgrades (Refined Weapons, Bolstered Plating, Optimized Energy Usage) for the Tal'darim Mothership.
Included a passive icon on the Tal'darim mothership describing how it has thermal lance/black hole installed baseline but doesn't have the base auxiliary systems baseline like the other two mothership Variants 
--- I wasn't sure if this passive icon was needed or not, I included for now but would be fine to remove it:
![image](https://github.com/user-attachments/assets/13d5783d-5ab2-4ab3-af52-74f5c0b78858)

Ziktofel commented on Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/EffectData.xml:
```diff
@@ -10961,7 +11030,7 @@
         <Target Value="CasterUnit"/>
     </CEffectIssueOrder>
     <CEffectApplyBehavior id="AP_ArbiterMPCloakingFieldApply">
-        <ValidatorArray index="0" value="NotMothership"/>
+        <ValidatorArray value="AP_NotMothershipAiur"/>
```
The vanilla validator targets vanilla Mothership, shall be kept

Ziktofel commented on Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/GameData/UpgradeData.xml:
```diff
@@ -17262,6 +17526,11 @@
         <!--  Mothership -->
         <EffectArray Reference="Unit,AP_SOAMothershipv4,ShieldRegenRate" Value="0.05"/>
         <EffectArray Reference="Unit,AP_MothershipTaldarim,ShieldRegenRate" Value="0.05"/>
+        <EffectArray Reference="Unit,AP_MothershipPurifier,ShieldRegenRate" Value="0.05"/>
+        <EffectArray Reference="Unit,AP_MothershipPurifierPowerSource,ShieldRegenRate" Value="0.05"/>
+        <EffectArray Reference="Unit,AP_MothershipCorePurifier,ShieldRegenRate" Value="0.05"/>
+        <EffectArray Reference="Unit,AP_MothershipCoreAiur,ShieldRegenRate" Value="0.05"/>
+        <EffectArray Reference="Unit,AP_MothershipAiur,ShieldRegenRate" Value="0.05"/>
```
Missing 100-stack filler for new stuff

Ziktofel commented:
Editor warnings:
```
[1/28/2025 8:39:15 PM] Warning: XML: Mods\ArchipelagoPlayer.SC2Mod\Base.SC2Data\GameData\EffectData.xml(11002,5) : Unable to create duplicate entry AP_MothershipCorePlasmaDisruptorLaunchMissile  Source: Mods/ArchipelagoPlayer.SC2Mod  Scope: CEffectLaunchMissile
[1/28/2025 8:39:15 PM] Warning: XML: Mods\ArchipelagoPlayer.SC2Mod\Base.SC2Data\GameData\EffectData.xml(18362,5) : Unable to create duplicate entry AP_MothershipCoreNexusRecallPrepare  Source: Mods/ArchipelagoPlayer.SC2Mod  Scope: CEffectSet
```

TODO:
- Tal'Darim weapon switch (See Infested Libs for that)
- Aux upgrades submenu need to be sticky (don't leave on upgrade purchase)
- Code cleanup (see review comments), closed those already solved
- Hotkey polishing


Other:
- The Motherships could start with appropriate core's abilities pre-installed (core's could be zero cost, also Taldarim's defaults)
- Auxilliary upgrades: an item per type (for all Motherships) or per-upgrade type (for all Motherships) like `Refined Weapons (all Motherships)`

Are those upgrades supposed to be War council? If not, the icon should be hidden (not greyed-out) instead. 
![Snímek obrazovky pořízený 2025-01-28 21-43-43](https://github.com/user-attachments/assets/3b5e1dce-4d80-4c57-af22-bec5a047fc8c)

![Snímek obrazovky pořízený 2025-01-28 21-43-48](https://github.com/user-attachments/assets/2ca9a36f-b9ac-4dc4-a448-1c8239fdffaa)

SirChuckOfTheChuckles commented:
Added Weapon Toggle To Tal'darim Mothership (configurable in Auxiliary Menu once researched)
![image](https://github.com/user-attachments/assets/21e94eb9-6882-40dd-9492-b576852a1d94)


Cleaned up stray code symbols and race-tinted some icons
Added 100 stack filler to shield regen
Made some hotkeys more convenient
Made Auxiliary Submenu sticky
Allowed Aiur and Purifier Motherships to install their mothership core abilities as Offensive Abilities (not installed automatically, but does come baseline at a cheap cost). 

The two greyed-out Mothership Core passives are intended to be War-council, to provide a more balanced unit when nerfed.

To address this:

"Auxilliary upgrades: an item per type (for all Motherships) or per-upgrade type (for all Motherships) like Refined Weapons (all Motherships)"

I want to not have the Motherships all share upgrades for now because I want to have separate costs for the stat upgrades for each Mothership to act as a tuning knob; I don't want Aiur Mothership weapon upgrade to cost as much as the Purifier Mothership since Purifier has more attacks baseline, for example. If it ends up being a problem on beta tests before live, I can change the item that gives Auxiliary upgrade to Tal'darim Mothership to have it give the stat auxiliaries to ALL motherships, and then have them share upgrades.

MatthewMarinets commented on Mods/ArchipelagoPlayer.SC2Mod/enUS.SC2Data/LocalizedData/GameHotkeys.txt:
```diff
@@ -620,4 +643,6 @@ Button/Hotkey/AP_ScoutTaldarimVulcanBlasterCancel=Escape
\ No newline at end of file
+Button/Hotkey/AP_ProbeWarpin=R
+Button/Hotkey/AP_MothershipTaldarimBaseAttackEnable=F
+Button/Hotkey/AP_MothershipTaldarimBaseAttackDisable=G
```
Note these aren't alphabetical and will thus be scrambled / generate a diff next time someone saves in the editor.

SirChuckOfTheChuckles replied:
Fixed, thanks for the heads up
